### ISSUE TITLE
Unable to read entire parquet directory

### ISSUE THREAD

**Issue by ddutt** [modin-project/modin#624]
<!--
General questions should be asked on the mailing list modin-dev@googlegroups.com.

Before submitting an issue, please fill out the following form.
-->

### System information
- Ubuntu 19.04
- pipenv install modin:
- **Modin version**: 0.5.0
- **Python version**: 3.7.3
- **Exact command to reproduce**: pd.read_parquet('parquet-out/foo/')

<!--
You can obtain the Modin version with

python -c "import modin; print(modin.__version__)"
-->

### Describe the problem
<!-- Describe the problem clearly here. -->
It seems that read_parquet cannot read an entire directory of parquet files i.e. partitioned parquet directory. 

### Source code / logs
<!-- Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem. -->

~/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/modin/pandas/io.py in read_parquet(path, engine, columns, **kwargs)                                                                         
     27     return DataFrame(                                                                                                                                                                                      
     28         query_compiler=BaseFactory.read_parquet(                                                                                                                                                           
---> 29             path=path, columns=columns, engine=engine, **kwargs                                                                                                                                            
     30         )                                                                                                                                                                                                  
     31     )                                                                                                                                                                                                      
                                                                                                                                                                                                                   
~/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/modin/data_management/factories.py in read_parquet(cls, **kwargs)                                                                           
     49     @classmethod                                                                                                                                                                                           
     50     def read_parquet(cls, **kwargs):                                                                                                                                                                       
---> 51         return cls._determine_engine()._read_parquet(**kwargs)                                                                                                                                             
     52                                                                                                                                                                                                            
     53     @classmethod                                                                                                                                                                                           
                                                                                                                                                                                                                   
~/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/modin/data_management/factories.py in _read_parquet(cls, **kwargs)                                                                          
     53     @classmethod                                                                                                                                                                                           
     54     def _read_parquet(cls, **kwargs):                                                                                                                                                                      
---> 55         return cls.io_cls.read_parquet(**kwargs)                                                                                                                                                           
     56                                                                                                                                                                                                            
     57     @classmethod                                                                                                                                                                                           
                                                                                                                                                                                                                   
~/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/modin/engines/ray/generic/io.py in read_parquet(cls, path, engine, columns, **kwargs)                                                       
    149                                                                                                                                                                                                            
    150         if not columns:                                                                                                                                                                                   
--> 151             pf = ParquetFile(path, memory_map=False)                                                                                                                                                      
    152             columns = [                                                                                                                                                                                   
    153                 name                                                                                                                                              
                                                                                                                                                                                                                  
~/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py in __init__(self, source, metadata, common_metadata, memory_map)
    127                  memory_map=True):                                                                                                                                                                        
    128         self.reader = ParquetReader()                                                                                                                                                                     
--> 129         self.reader.open(source, use_memory_map=memory_map, metadata=metadata)                                                                                                                            
    130         self.common_metadata = common_metadata                                                                                                                                                             
    131         self._nested_paths_by_prefix = self._build_nested_paths()

~/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/_parquet.pyx in pyarrow._parquet.ParquetReader.open()

~/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/error.pxi in pyarrow.lib.check_status()
ArrowIOError: Arrow error: IOError: Error reading bytes from file: Is a directory

**Comment by devin-petersohn** [comment]
Hi @ddutt, thanks for reporting!

How was this directory created? Is this directory working to read into pandas?

From my understanding, this wouldn't work in pandas, so it could be a good addition to our `experimental` API. The goal of that API is to make clear what are "augmentations" on top of the pandas API where the behavior diverges.

**Comment by ddutt** [comment]
Hi,

Appreciate the prompt response.

> Hi @ddutt, thanks for reporting!
> 
> How was this directory created? Is this directory working to read into pandas?

 Yes, the directory can be read in Pandas. 

> 
> From my understanding, this wouldn't work in pandas, so it could be a good addition to our `experimental` API. The goal of that API is to make clear what are "augmentations" on top of the pandas API where the behavior diverges.

Nope, it does work in pandas correctly. Showing the ipython snippet below:

In [5]: df = pd.read_parquet('parquet-out/foo')

In [6]: df.info()
<class 'pandas.core.frame.DataFrame'>
Int64Index: 454 entries, 1 to 4
Data columns (total 20 columns):
...

And an ls to show that there is more dir under parquet-out/foo:

$ ls -l parquet-out/foo/
total 56
drwxrwxr-x 13 ddutt ddutt 4096 May  5 09:39 a1
drwxrwxr-x 24 ddutt ddutt 4096 May  5 09:39 a2
drwxrwxr-x 24 ddutt ddutt 4096 May  5 09:39 a3
drwxrwxr-x  7 ddutt ddutt 4096 May  5 09:39 b1
drwxrwxr-x 30 ddutt ddutt 4096 May  5 09:39 b2
drwxrwxr-x 30 ddutt ddutt 4096 May  5 09:39 b3
drwxrwxr-x 30 ddutt ddutt 4096 May  5 09:39 b4
drwxrwxr-x 30 ddutt ddutt 4096 May  5 09:39 b5
drwxrwxr-x  9 ddutt ddutt 4096 May  5 09:39 c1
drwxrwxr-x  9 ddutt ddutt 4096 May  5 09:39 d1
drwxrwxr-x  9 ddutt ddutt 4096 May  5 09:39 d2
drwxrwxr-x  9 ddutt ddutt 4096 May  5 09:39 d3
drwxrwxr-x 14 ddutt ddutt 4096 May  5 09:39 e1
drwxrwxr-x 14 ddutt ddutt 4096 May  5 09:39 e2

Please let me know if you need further info from me. I can have many parquet files and so I'd very much like the ability to read in parallel an entire directory. Alternately, is there any other way to read the files in parallel?

Thanks for providing this software in open source. 

Dinesh

**Comment by ddutt** [comment]
Meanwhile, I found the workaround as: pa.ParquetDataSet(folder).read().to_pandas().

Dinesh

**Comment by devin-petersohn** [comment]
Thanks @ddutt, sorry I am traveling today and have limited WiFi until this evening.

We are using parquet in the partitions, so it should be a quick fix. Link to relevant code here: https://github.com/modin-project/modin/blob/master/modin/engines/ray/pandas_on_ray/io.py#L54

**Comment by williamma12** [comment]
Hi @ddutt, I was unable to reproduce the error of what you had exactly. I got a `ArrowIOError: Memory mapping file failed: Invalid argument` and it was resolved with #632. Could try again with #632 and let me know if that fixes your issue? Thanks!

**Comment by ddutt** [comment]
Hi, I left a comment on the other thread. If I add just your fixes to the files I got via pip install modin, I get a different error. I've turned off memory map and so the problem you have doesn't arise. What I get instead is this new error:
RayTaskError: ray_worker (pid=31117, host=ddutt-yoga)
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/modin/engines/ray/pandas_on_ray/io.py", line 55, in _read_parquet_columns                                                
    df = pq.read_pandas(path, columns=columns, **kwargs).to_pandas()
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py", line 1144, in read_pandas                                                         
    use_pandas_metadata=True)
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py", line 1123, in read_table                                                          
    use_pandas_metadata=use_pandas_metadata)
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/filesystem.py", line 181, in read_parquet                                                      
    use_pandas_metadata=use_pandas_metadata)
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py", line 985, in read                                                                 
    use_pandas_metadata=use_pandas_metadata)
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py", line 535, in read                                                                 
    table = reader.read(**options)
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py", line 211, in read                                                                 
    columns, use_pandas_metadata=use_pandas_metadata)
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py", line 261, in _get_column_indices                                                  
    for descr in index_columns
  File "/home/ddutt/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py", line 262, in <listcomp>                                                           
    if descr['kind'] == 'serialized']
TypeError: string indices must be integers

**Comment by williamma12** [comment]
@ddutt, I am unable to recreate your new issue. It works for me when I write the parquet files using 
```
import pyarrow.parquet as pq
import pyarrow as pa
table = pa.Table.from_pandas(pandas_df)
pq.write_to_dataset(table, root_path='test')
```
or 
```
import pandas
pandas_df.to_parquet('test', engine='pyarrow', partition_cols=['col1'])
```
The second method is causing some issues similar to #633 that I am currently fixing.

For reference, 
```
import pandas
pandas_df = pandas.DataFrame({"col1": [1,2,3], "col2": [4,5,6], "col3": [7, 8, 9]})
```

Could you provide me with how you are writing your parquet files and some sample data (including columns and index)?

**Comment by ddutt** [comment]
Here is the code I use to write the parquet files:
`
        table = pa.Table.from_pandas(df, schema=data['schema'],
                                                           preserve_index=False)

        pq.write_to_dataset(table, root_path=cdir,
                            partition_cols=data['partition_cols'],
                            version="2.0",
                            flavor='spark')`

**Comment by williamma12** [comment]
@ddutt sorry for the multiple fixes but I am still unable to recreate the `TypeError: string indices must be integers` error. However, #632 might now fix your error. The previous fix did not handle `partition_cols` properly, but now #632 adds support for handling partitioned parquet files with specified `partition_cols`.

**Comment by ddutt** [comment]
I'll take out the version and flavor options from my parquet write and see if that fixes the problem. Nevertheless, its surprising that pandas has no problems reading but modin does.

**Comment by ddutt** [comment]
Nope, same problem.

**Comment by ddutt** [comment]
I even took out the preserve_index=False option. No joy.

**Comment by williamma12** [comment]
hmm ok i'll look more into it later tonight. sorry for the delay!

**Comment by devin-petersohn** [comment]
Hi @ddutt, sorry I've been away from the thread. I'm at a conference this week.

Is this data you can share? Since we're having such difficulty reproducing the error, it seems like sharing the file is the only way we can fix it.

**Comment by ddutt** [comment]
I believe I can. Let me see what I can do. Thanks again to both of you for all your help. One last q. I've been writing with pyarrow not from the ray distribution, but from the original pyarrow distribution. Do you think that could be a cause. I can try that option as well.

**Comment by williamma12** [comment]
I think I tried it with pyarrow from the ray distribution and from the pip installed version and there was no difference. Though I will try it again tonight and try writing the data from the spark version as well

**Comment by ddutt** [comment]
OK, when I tried to write with the pyarrow version from ray (by adding import modin.pandas as pd first), my write fails with:

Traceback (most recent call last):
  File "q.py", line 166, in <module>
    _main(userargs)
  File "q.py", line 114, in _main
    loop.run_until_complete(asyncio.gather(*tasks))
  File "/usr/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
  File "writer.py", line 94, in run_output_worker
    worker.write_data(data)
  File "writer.py", line 53, in write_data
    table = pa.Table.from_pandas(df, schema=data['schema'])
  File "pyarrow/table.pxi", line 1139, in pyarrow.lib.Table.from_pandas
  File "/home/ddutt/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/pandas_compat.py", line 471, in dataframe_to_arrays                                            
    convert_types)]
  File "/home/ddutt/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/pandas_compat.py", line 470, in <listcomp>                                                     
    for c, t in zip(columns_to_convert,
  File "/home/ddutt/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/pandas_compat.py", line 466, in convert_column                                                 
    raise e
  File "/home/ddutt/.local/share/virtualenvs/foo-cFmMv4Vf/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/pandas_compat.py", line 460, in convert_column                                                 
    return pa.array(col, type=ty, from_pandas=True, safe=safe)
  File "pyarrow/array.pxi", line 173, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 36, in pyarrow.lib._sequence_to_array
  File "pyarrow/error.pxi", line 81, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: ('Could not convert True with type numpy.bool_: tried to convert to boolean', 'Conversion failed for column active with type bool')

**Comment by ddutt** [comment]
So, at first glance it seems the problem has to do with the pyarrow version in ray.

**Comment by devin-petersohn** [comment]
@ddutt If it is the pyarrow version in ray, we can easily fix that. Do you have some sample data (not necessarily this data + the exact call) that can reproduce the issue?

Also what is your environment pyarrow version?

**Comment by ddutt** [comment]
pyarrow version is 0.12.1.

I'll be back a bit later to try and get you the data.

**Comment by devin-petersohn** [comment]
Would you be able to try it with 0.13? There may have been some parquet updates that broke some backwards compatibility.

**Comment by ddutt** [comment]
I switched to pyarrow 0.13 and now I get a different error:
```
In [3]: df                          
Out[3]: ---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)                     
~/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)                                                                                                                                                
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    400                         if cls is not object \
    401                                 and callable(cls.__dict__.get('__repr__')):
--> 402                             return _repr_pprint(obj, self, cycle)
    403
    404             return _default_pprint(obj, self, cycle)

~/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    695     """A pprint that just redirects to the normal repr function."""
    696     # Find newlines and replace them with p.break_()
--> 697     output = repr(obj)
    698     for idx,output_line in enumerate(output.splitlines()):
    699         if idx:

~/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/modin/pandas/dataframe.py in __repr__(self)
     89         num_cols = pandas.get_option("max_columns") or 20
     90
---> 91         result = repr(self._build_repr_df(num_rows, num_cols))
     92         if len(self.index) > num_rows or len(self.columns) > num_cols:
     93             # The split here is so that we don't repr pandas row lengths.

~/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/modin/pandas/base.py in _build_repr_df(self, num_rows, num_cols)
     48
     49         if not hasattr(self, "columns") or len(self.columns) <= num_cols:
---> 50             head_front = head.to_pandas()
     51             # Creating these empty to make the concat logic simpler
     52             head_back = pandas.DataFrame()

~/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/modin/backends/pandas/query_compiler.py in to_pandas(self)
    485         else:
    486             ErrorMessage.catch_bugs_and_request_email(
--> 487                 len(df.index) != len(self.index) or len(df.columns) != len(self.columns)
    488             )
    489             df.index = self.index

~/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/modin/error_message.py in catch_bugs_and_request_email(cls, failure_condition)
     36         if failure_condition:
     37             raise Exception(
---> 38                 "Internal Error. "
     39                 "Please email bug_reports@modin.org with the traceback and command that"
     40                 " caused this error."

Exception: Internal Error. Please email bug_reports@modin.org with the traceback and command that caused this error.
```

**Comment by ddutt** [comment]
In other words, the read_parquet error is gone. But I get a different error when I attempt to display the contents of the dataframe. I've hand patched the fixes @williamma12 provided.

**Comment by ddutt** [comment]
On debugging further, it seems that when modin splits the read of the partitioned parquet directory, it adds back the columns and names that are in the partitioned directory wrongly. For example, if I create a partitioned parquet read, and I have the structure topdir/var1=a/var2=b/foo.parquet, and the columns in foo.parquet are var3, var4, var5, the new columns after modin reads is like this: var3, var1, var2, var4, var1, var2, var5, var1, var2. Here is the debug statement that I see:

```
~/.local/share/virtualenvs/a-cFmMv4Vf/lib/python3.7/site-packages/modin/backends/pandas/query_compiler.py in to_pandas(self)
    485         else:
    486             ErrorMessage.catch_bugs_and_request_email(
--> 487                 len(df.index) != len(self.index) or len(df.columns) != len(self.columns)
    488             )
    489             df.index = self.index
ipdb> df.index
RangeIndex(start=0, stop=12, step=1)
ipdb> self.index
RangeIndex(start=0, stop=12, step=1)
ipdb> df.columns
Index(['var3', 'var1', 'var2', 'var4', 'var1', 'var2',
       'var5', 'var1', 'var2'], dtype='object')
ipdb> self.columns
Index(['var3', 'var4', 'var5'], dtype='object')
```

**Comment by williamma12** [comment]
Ah yes that should be fixed in the newest version of #632. Sorry about that I pushed some more commits earlier today to resolve that issue. You should be able to install it with 
```
pip install git+https://github.com/williamma12/modin.git@io/parquet_partitions
```

**Comment by ddutt** [comment]
YES! It works on applying the additional patch. 

Thank you so much @williamma12 for so diligently working on the problem. You can mark the bug fixed at least from the perspective of the way I reported it. 

Best wishes,

Dinesh

**Comment by williamma12** [comment]
Yay! Thanks for working so closely with us on getting this bug fixed @ddutt. We'll continue looking into what happened between ray's pyarrow, which is built off of 0.12.1, pyarrow 0.12.1, and pyarrow 0.13.

**Comment by devin-petersohn** [comment]
@pcmoritz Do you think we could update the version of Pyarrow shipped with Ray?

**Comment by pcmoritz** [comment]
The pyarrow version shipped with ray is now upgraded in https://github.com/ray-project/ray/pull/4858
